File: BasicML/basicml/optim/adagrad.py
from .optimizer             import Optimizer
from typing                 import Optional
from basicml.regularization import Regularizer
from basicml.tensor         import Tensor
import numpy as np

class Adagrad(Optimizer):
    def __init__(self,
                 parameters : list[Tensor],
                 lr         : float,
                 regularizer: Optional[Regularizer] = None,
                 eps        : float                 = 1e-8):
        super().__init__(parameters, lr, regularizer)
        self.eps            = eps
        self.accumulated_sq = [np.zeros_like(p.data) for p in self.parameters]

    def step(self):
        for i, param in enumerate(self.parameters):
            if param.requires_grad and param.grad is not None:
                grad = param.grad
                if self.regularizer is not None:
                    grad = grad + self.regularizer.grad(param)
                self.accumulated_sq[i] += grad ** 2
                param.data -= self.lr * grad / (np.sqrt(self.accumulated_sq[i]) + self.eps)

    def zero_grad(self):
        for param in self.parameters:
            param.zero_grad()

File: AdvanceML/src/optim/adagrad.cpp
#include "advanceml/optim/adagrad.hpp"

#include <cmath>

namespace advanceml {

Adagrad::Adagrad(std::vector<Tensor> parameters, float learning_rate, float eps)
    : parameters_(std::move(parameters)), learning_rate_(learning_rate), eps_(eps) {
    accumulated_sq_.reserve(parameters_.size());
    for (const Tensor& param : parameters_) {
        accumulated_sq_.emplace_back(param.data().size(), 0.0f);
    }
}

void Adagrad::step() {
    for (size_t i = 0; i < parameters_.size(); ++i) {
        Tensor& param = parameters_[i];
        if (!param.has_grad()) {
            continue;
        }
        Tensor grad = param.grad();
        FloatBuffer& values = param.mutable_data();
        std::vector<float>& accumulated_sq = accumulated_sq_[i];
        for (size_t j = 0; j < values.size(); ++j) {
            accumulated_sq[j] += grad.data()[j] * grad.data()[j];
            values[j] -= learning_rate_ * grad.data()[j] / (std::sqrt(accumulated_sq[j]) + eps_);
        }
    }
}

void Adagrad::zero_grad() {
    for (Tensor& param : parameters_) {
        param.zero_grad();
    }
}

}  // namespace advanceml
